app/controllers/api/stages_controller.rb

```diff
@@ -0,0 +1,42 @@
+class Api::StagesController < Api::BaseController
+  include CurrentProject
+
+  before_action :require_project
+
+  def index
+    render json: @project.stages
+  end
+
+  def clone
+    new_stage = Stage.build_clone(old_stage)
+    new_stage.name = stage_name
+    if new_stage.save
+      if deploy_group_ids.any? && deploy_groups.any?
+        new_stage.deploy_groups = deploy_groups
+        new_stage.save
+      end
+
+      render json: new_stage, status: 201
+    else
+      render json: new_stage.errors, status: 422
```

Comment:
I think we should `save!` everywhere and then rescue the invalid error to present the error ...